### PR TITLE
Add renaming support

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
I'll add the enforced fields renaming in the same PR. Running a few tests in the meantime. :)

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Should be complete now. Just wondering about the formatting stuff...

File: src/clone.rs [hunk 1]
```diff
@@ -163,6 +179,21 @@ macro_rules! to_return_value {
...
 ///
+/// ### Renaming variables
+///
+/// ```
+/// use glib::clone;
+/// use std::rc::Rc;
+///
+/// let v = Rc::new(1);
+/// let u = Rc::new(2);
+/// let closure = clone!(@strong v as y, @weak u => move |x| {
```
Comment (EPashkin) [inline src/clone.rs]: @GuillaumeGomez  can you add test with two @strong renamed variable too?

Comment (GuillaumeGomez) [reply src/clone.rs]: What would it change?

Comment (EPashkin) [reply src/clone.rs]: Just to show what syntax used in this case

File: src/clone.rs [hunk 2]
```diff
@@ -246,63 +244,131 @@ macro_rules! to_return_value {
...
+/// # use std::rc::Rc;
+/// #[derive(Debug)]
+/// struct Foo {
+///     v: Rc<usize>,
+/// }
+///
+/// impl Foo {
+///     fn foo(&self) {
+///         let closure = clone!(@strong self.v => move |x| {
+///             println!("v: {:?}", v);
```
Comment (sdroege) [inline src/clone.rs]: I don't think we should allow that. It allows annoying bugs to sneak in

```rust
clone!(@strong self.v, self.bla.v => move |x| {
    // what's `v` here now? Might even be the same type so the closure would silently misbehave
});
```

I believe we should require renaming for `self` and field accesses and not come up with any magic names ourselves.

File: tests/clone.rs [hunk 3]
```diff
@@ -36,6 +36,27 @@ fn clone_and_references() {
...
+fn subfields_renaming() {
+    struct Foo {
+        v: Rc<usize>,
+    }
+
+    impl Foo {
+        fn foo(&self) {
+            let state = Rc::new(RefCell::new(State::new()));
+
+            let closure = clone!(@strong self.v as v, @weak state as hello => move |_| {
```
Comment (sdroege) [inline tests/clone.rs]: Can you also add a test for `self` renaming?

File: src/clone.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/clone.rs]: We don't. The block is marked as "compile_fail" you know. :p

File: tests/clone.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply tests/clone.rs]: Already done in the docs.

File: src/clone.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/clone.rs]: I'm not sure to follow...

Comment (EPashkin) [reply src/clone.rs]: I was just not sure if this case work or need two `@strong`, now you can ignore this comment
```
clone!(@strong a as b, c as d => ...)
```

Comment (GuillaumeGomez) [reply src/clone.rs]: The example you wrote doesn't work either: you have to add `@strong` or `@weak` before each variable.

File: src/clone.rs (hunk 2 above)
Comment (sdroege) [reply src/clone.rs]: Ah please remove this from the docs then. People like me assume that examples in the docs actually work :) I didn't see that annotation

File: tests/clone.rs (hunk 3 above)
Comment (sdroege) [reply tests/clone.rs]: Ok!

File: src/clone.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/clone.rs]: People reading the docs will see the "this code doesn't compile" element. rustdoc stuff you know. :p

Comment (sdroege) [reply src/clone.rs]: How does that look like in the docs, can you show an example in e.g. the std docs? It seems strange to me to include broken code examples in the docs.

Comment (sdroege) [reply src/clone.rs]: Ok you convinced me

File: src/clone.rs (hunk 1 above)
Comment (sdroege) [reply src/clone.rs]: @EPashkin anything missing here?

Comment (EPashkin) [reply src/clone.rs]: IMHO all good

Comment (EPashkin) [reply src/clone.rs]: except for new merge conflict 😢

**Comment by GuillaumeGomez** [comment]
Urg. Where is this conflict coming from?! ><

**Comment by fengalin** [comment]
@GuillaumeGomez: sorry, my fault, I [ran `‘fmt`](https://github.com/gtk-rs/glib/commit/410c4ae5dedc747e8898c3dbf28241646ee5ec62).

**Comment by GuillaumeGomez** [comment]
Fixed the conflict. Let's see if the CI is still happy.

**Comment by GuillaumeGomez** [comment]
@fengalin `fmt` is definitely my personal nemesis...

**Pull request merged**